src/port1.0/portlint.tcl

```diff
@@ -115,6 +116,132 @@ proc portlint::seems_utf8 {str} {
     return true
 }
 
+# lint_checksum_types
+#
+# Given a list of checksum types, return a list of strings which are warnings
+# about deprecated checksum types, or missing recommended types.
+#
+# Returns an empty list if no issues are found.
+proc portlint::lint_checksum_type_list {types} {
+    set issues [list]
+
+    foreach type $types {
+        if {$type ni [portchecksum::list_preferred_checksum_types]} {
+            lappend issues "checksum type is deprecated: $type"
+        }
+    }
+
+    foreach preferred [portchecksum::list_preferred_checksum_types] {
+        if {$preferred ni $types} {
+            lappend issues "missing recommended checksum type: $preferred"
+        }
+    }
+
+    return $issues
+}
+
+# lint_checksum
+#
+# Checks a given Portfile checksum string.  Returns a list of lists.
+# The first member list is a list of error strings.
+# The second member list is a list of warning strings.
+#
+# Returns a list containing two empty lists if no issues are found.
+proc portlint::lint_checksum {checksum_string} {
+    set errors [list]
+    set warnings [list]
+
+    set is_error false
+    set ctr_start 0
+
+    set filename ""
+    set pfx ""
+    set has_filenames false
+
+    set types [list]
+
+    # List of all tokens in the checksum string
+    set checksum_tokens [regexp -all -inline {\S+} $checksum_string]
+
+    if {[string equal [lindex $checksum_tokens 0] "checksum"]} {
+        incr ctr_start
+    }
+
+    for {set ctr $ctr_start} \
+        {[expr $ctr < [llength $checksum_tokens]] && !$is_error} \
+        {} {
+
+        set current [lindex $checksum_tokens $ctr]
+
+        if {$current in [portchecksum::list_checksum_types]} {
+            set c_type  $current
+            set c_value [lindex $checksum_tokens [expr $ctr + 1]]
```

Comment:
Input to expr should be passed as a single argument, wrapped in braces. http://wiki.tcl.tk/10225

Also, lindex can do simple arithmetic directly, so this expr call isn't needed.